File: Haskell/euler010/src/Euler010.hs
-- Project Euler - 010 - Summation of primes
--
-- The sum of the primes below 10 is 2 + 3 + 5 + 7 = 17.
-- Find the sum of all the primes below two million.
--
-- NOTE:
-- Prime function from Euler 007
module Euler010 (euler010, sumPrimesToN, primesToN) where

import Data.List.Ordered (minus)

-- Unbounded natural numbers are about 0.1s slower for the 2,000,000th prime
-- import Numeric.Natural
-- type Nat = Natural

type Nat = Word

euler010 :: Nat
euler010 = sumPrimesToN 2000000

sumPrimesToN :: Nat -> Nat
sumPrimesToN = sum . primesToN

primesToN :: Nat -> [Nat]
primesToN n = doRecursive sieve [2..n]
  where
    sieve x xs = xs `minus` [x*x, x*x+x..n]

-- |doRecursive: Apply a function which operates on a list recursively to the
-- |tail of a list, keeping the head each time.
doRecursive :: (a -> [a] -> [a]) -> [a] -> [a]
doRecursive f []     = []
doRecursive f (x:xs) = x : doRecursive f (f x xs)

File: CXX/src/euler010.cpp
/*
 * Project Euler - 010 - Summation of primes
 * Tamara Roberson <tamara.roberson@gmail.com>
 * Copyright (c) 2019 Tamara Roberson
 *
 * The sum of the primes below 10 is 2 + 3 + 5 + 7 = 17.
 * Find the sum of all the primes below two million.
 */

#include <euler/euler010.hpp>
#include <euler/utils/primes.hpp> // utils::primes_up_to
#include <euler/utils/sum.hpp> // utils::sum

#include <cstdint> // int64_t
#include <vector> // std::vector

namespace euler
{

int euler010_example()
{
    auto primes = utils::primes_up_to<int>(10);
    return utils::sum<int>(primes);
}

int64_t euler010()
{
    auto primes = utils::primes_up_to<int64_t>(2000000);
    return utils::sum<int64_t>(primes);
}

} // end namespace euler
